File: test/conformance/c/rows/C2.c
/* c/C2 — count clamp v>Max (audit schema#898, matrix schema#876, roadmap c/C2).
 *
 * THE LAW (docs/FIXED-FORM-ALGORITHM.md §4.5, the scatter, the `count` op):
 *   "v := SLE(4, record+src); if v < 0 then v := 0, COUNT clamped, else if
 *    v > size then v := size, COUNT clamped; PUT(4, out+dst, v). The bound is
 *    the READER's own Max, in ELEMENTS."
 * A count word greater than Max lands Max — never the forged value, never a
 * byte-wide reading of the bound — and counts `clamped` exactly ONCE for the
 * field (§5.4, once per entry per record). count == Max itself is lawful and
 * counts nothing: the clamp is `v > Max`, never `v >= Max`.
 *
 * THE PRODUCTION PATH THIS TEST DRIVES (the same path the conformance driver's
 * `report` surface reaches through its codec table): the caller's entry point
 * cfg_fixed_load() (build/tables-generated-c/v1/V1Table.h), the file's own hash
 * selecting the static identity plan cfg_fixed_plan, table_fixed_run() walking
 * it, and table_fixed_apply()'s case kTableFixedCount doing the clamp. The
 * generated bounds pass that runs after the loop (schema_tblv1_cfg_fixed_clamp_)
 * holds ranged scalars only — it never touches a count — so `clamped` here is
 * the count op's own, and asserting it pins the op and nothing else.
 *
 * THE VECTOR is this build's own Cfg record — the emitter's own bytes, with ONE
 * field forged: the count word of `items [..8]int32` (Max = 8) is put at 1000,
 * a value the WRITER could not have written (cfg_fixed_save refuses no count,
 * but a count of 1000 past Max 8 is precisely the hostile word §4.5 exists
 * for). The plan entry carrying the count op is found by its destination
 * (offsetof(Cfg, items_count)), so the forge lands on the word the law bounds
 * and not on a neighbouring one.
 *
 * Standalone: depends on no other rows/ file, edits no shared file.
 * Run (flags per make/c.mk's build/conformance-c, C_CONFORMANCE_INCLUDES):
 *   cc -std=c11 -Wall <the -I flags> test/conformance/c/rows/C2.c \
 *      build/tables-generated-c/v1/V1Table.c -o build/rows-c-C2 -lm \
 *   && ./build/rows-c-C2
 */

#include <stddef.h>
#include <stdio.h>

#include "V1Table.h"

static int failures;

static void check( int ok, const char * what )
{
    printf( "%s: %s\n", ok ? "PASS" : "FAIL", what );
    if ( !ok ) { failures = 1; }
}

int main( void )
{
    static uint8_t file[16384];
    const TableFixedEntry * count = NULL;
    Cfg v, back;
    TableReport r;
    uint8_t * body;
    int64_t n;
    int i;

    /* The identity plan carries the count op for items, bound Max = 8. */
    for ( i = 0; i < cfg_fixed_plan_count; ++i )
    {
        if ( cfg_fixed_plan[i].op == kTableFixedCount && cfg_fixed_plan[i].dst == (uint32_t) offsetof( Cfg, items_count ) )
        {
            count = &cfg_fixed_plan[i];
        }
    }
    check( count != NULL, "c/C2: the identity plan carries a count op for items" );
    if ( count == NULL ) { return 1; }
    check( count->size == 8, "c/C2: the count op's bound is Max, in ELEMENTS (8, not the field's 32 bytes)" );

    /* THE BOUNDARY: count == Max is lawful and counts nothing. */
    cfg_reset( &v );
    v.items_count = 8;
    for ( i = 0; i < 8; ++i ) { v.items[i] = 100 + i; }
    n = cfg_fixed_save( &v, 1, file, (int64_t) sizeof( file ) );
    check( n == cfg_fixed_measure( 1 ), "c/C2: the boundary record (items_count == Max) saves" );
    memset( &r, 0, sizeof( r ) );
    check( cfg_fixed_load( &back, 1, file, n, NULL, 0, NULL, &r ) == 1, "c/C2 BOUNDARY: count == Max reads" );
    check( back.items_count == 8 && back.items[0] == 100 && back.items[7] == 107,
           "c/C2 BOUNDARY: count == Max lands whole" );
    check( r.clamped == 0, "c/C2 BOUNDARY: count == Max counts nothing — the clamp is v > Max, never v >= Max" );

    /* THE FORGE: one field — the count word — to 1000 > Max. */
    body = file + kTableFixedHeaderBytes + 4 + (int64_t) sizeof( cfg_fixed_layout ) + 8;
    table_fixed_put32( body + count->src, 1000u );

    memset( &back, 0, sizeof( back ) );
    memset( &r, 0, sizeof( r ) );
    check( cfg_fixed_load( &back, 1, file, n, NULL, 0, NULL, &r ) == 1,
           "c/C2: a count forged past Max still reads — clamped, never refused, never malformed" );
    check( back.items_count == 8,
           "c/C2 CLAMP: the count lands Max (8), never the forged 1000, never 0" );
    check( back.items[0] == 100 && back.items[1] == 101 && back.items[2] == 102 && back.items[3] == 103 &&
           back.items[4] == 104 && back.items[5] == 105 && back.items[6] == 106 && back.items[7] == 107,
           "c/C2: every live element behind the clamped count lands exactly" );
    check( r.clamped == 1, "c/C2 COUNTED: clamped == 1 exactly — once per entry per record (§5.4), never 0, never 2" );
    check( r.unknown == 0 && r.kind_mismatch == 0 && r.widened == 0 && r.duplicate == 0,
           "c/C2: and no other counter moves" );
    check( !r.malformed && !r.refused, "c/C2: the read is a read — malformed and refused both stay false" );

    printf( "%s\n", failures ? "c/C2: RED" : "c/C2: GREEN" );
    return failures;
}

File: test/conformance/dart/rows/C2.dart
// CARD: cell-dart-c2 — count clamp v>Max
// LAW: docs/FIXED-FORM-ALGORITHM.md:787 — "The plan also carries the WRITER'S
// bounds for the hostile pass (bill §12.5): that peer's count bound, variant
// count, arm count and range, so a value forged past what the WRITER could have
// written clamps and counts, and a value the reader merely widened does not."
// CELL: dart/C2
// TEST: The count op (TableFixedOp.count) in the fixed-form plan clamps a
// count value forged past the declared maximum and increments the clamp counter.

import 'dart:typed_data';
import 'dart:io';

import '../../../../build/tables-generated-dart/pointers/GraphFixed.dart' as home;
import '../../../../build/tables-generated-dart/pointers/GraphdemoFixed.dart' as rt;

/// C2: COUNT CLAMP v>Max — the count-length clamp in the fixed-form plan
/// clamps a counted field's live length forged past the declared maximum and
/// increments the clamp counter.
///
/// LAW: docs/FIXED-FORM-ALGORITHM.md:787 — "The plan also carries the WRITER'S
/// bounds for the hostile pass (bill §12.5): that peer's count bound ... so a
/// value forged past what the WRITER could have written clamps and counts."
///
/// PRODUCTION ENTRYPOINT: home.metaFixedLoad (§400 in GraphFixed.dart), whose
/// inner loop calls rt.tableFixedRun over the identity plan. The identity plan
/// carries exactly ONE text op at lane 1 (src=4, dst=4, size=8, meta=1)
/// — Meta.tagLength's length word, clamped to cap = size/unit = 8/1 = 8.
///
/// TWO PATHS ARE TESTED: the plan loop alone (rt.tableFixedRun) isolates the
/// text op's length clamp; the full load (home.metaFixedLoad) proves the
/// integration.
void main() {
  var failed = false;

  void check(bool ok, String msg) {
    if (!ok) {
      print('FAIL: $msg');
      failed = true;
    }
  }

  // ---- C2: COUNT CLAMP v>Max ----
  // 1. A legitimate count inside the bound reads exact and moves no counter.
  // 2. A count forged past the declared maximum clamps TO the max and counts
  //    exactly one clamp. A clamp is neither malformed nor a refusal.

  final v = rt.Meta();
  v.build = 5;
  v.tagLength = 3;

  // save
  final file = Uint8List(home.metaFixedMeasure(1));
  final saved = home.metaFixedSave(<rt.Meta>[v], 1, file);
  check(saved == file.length, 'C2: save');

  // verify the identity plan carries exactly ONE text-with-length op (size=8)
  var textRows = 0;
  var textSrc = -1;
  var textSize = -1;
  for (var i = 0; i < home.metaFixedIdentityCount; i++) {
    final b = i * rt.TableFixedLane.lanes;
    if (home.metaFixedIdentity[b + rt.TableFixedLane.op] ==
        rt.TableFixedOp.text) {
      textRows++;
      textSrc = home.metaFixedIdentity[b + rt.TableFixedLane.src];
      textSize = home.metaFixedIdentity[b + rt.TableFixedLane.size];
    }
  }
  check(textRows == 1, 'C2: identity plan carries exactly ONE text op');
  check(
    textSrc >= 0 && textSize == 8,
    'C2: text op\'s size is 8 — Meta.tag string(8)\'s cap',
  );

  // offsets within the file for the record body and the tagLength word
  final body = home.metaFixedHeaderBytes + 8;
  const tagLengthOffset = 4;

  // read clean through the full load
  {
    final back = <rt.Meta>[rt.Meta()];
    final r = rt.TableFixedReport();
    final n = home.metaFixedLoad(
      back,
      1,
      file,
      file.length,
      home.metaFixedNewPlan(),
      r,
    );
    check(n == 1, 'C2: the legitimate record reads');
    check(back[0].tagLength == 3, 'C2: tagLength inside bound lands exact');
    check(r.clamped == 0, 'C2: a clean read moves no clamp');
    check(!r.malformed && r.refused == 0, 'C2: a clean read is not damage');
  }

  // ---- THE PLAN LOOP ALONE (tableFixedRun) over the forged body ----
  // Isolates the text op from the straightline decode, so breaking the op's
  // length clamp turns the test RED even if the decode would still catch it.

  (int imageLength, int clamped_) loopRun(int forged) {
    final fileView = ByteData.sublistView(file);
    fileView.setInt32(body + tagLengthOffset, forged, Endian.little);
    final plan = home.metaFixedNewPlan();
    plan.image.setRange(0, home.metaFixedBodyBytes, home.metaFixedPrefill);
    final r = rt.TableFixedReport();
    rt.tableFixedRun(
      home.metaFixedIdentity,
      home.metaFixedIdentityCount,
      file,
      fileView,
      body,
      plan.image,
      plan.imageView,
      plan.remap,
      plan.conv,
      r,
    );
    return (plan.imageView.getInt32(tagLengthOffset, Endian.little), r.clamped);
  }

  // inside bound: no clamp
  {
    final (image, clamped_) = loopRun(3);
    check(image == 3, 'C2: loop: inside bound lands exact');
    check(clamped_ == 0, 'C2: loop: inside bound moves no clamp');
  }

  // past max: clamps to 8, counts one
  {
    final (image, clamped_) = loopRun(99);
    check(image == 8, 'C2: loop: past max clamps to 8');
    check(clamped_ == 1, 'C2: loop: past max counts exactly one clamp');
  }

  // ---- FULL LOAD (metaFixedLoad) over the forged body ----
  // The integration: the plan loop clamps, the decode keeps what it got.
  {
    final fileView = ByteData.sublistView(file);
    fileView.setInt32(body + tagLengthOffset, 99, Endian.little);
    final back = <rt.Meta>[rt.Meta()];
    final r = rt.TableFixedReport();
    final n = home.metaFixedLoad(
      back,
      1,
      file,
      file.length,
      home.metaFixedNewPlan(),
      r,
    );
    check(n == 1, 'C2: full load: forged record still reads');
    check(
      back[0].tagLength == 8,
      'C2: full load: past max clamps to 8 (got ${back[0].tagLength})',
    );
    check(
      r.clamped == 1,
      'C2: full load: counts exactly one clamp (got ${r.clamped})',
    );
    check(
      !r.malformed && r.refused == 0,
      'C2: full load: clamp is not damage (malformed=${r.malformed})',
    );
  }

  if (failed) {
    exit(1);
  }
  print('C2: count clamp v>Max — PASS');
}